src/libmongoc/tests/test-mongoc-shared.c

```diff
@@ -0,0 +1,86 @@
+
+
+#include "./TestSuite.h"
+
+#include <mongoc/mongoc-shared-private.h>
+
+typedef struct {
+   int value;
+   int *store_value_on_dtor;
+} my_value;
+
+my_value *
+my_value_new ()
+{
+   my_value *p = bson_malloc0 (sizeof (my_value));
+   p->value = 42;
+   p->store_value_on_dtor = NULL;
+   return p;
+}
+
+void
+my_value_free (my_value *ptr)
+{
+   if (ptr->store_value_on_dtor) {
+      *ptr->store_value_on_dtor = ptr->value;
+   }
+   ptr->value = 0;
+   ptr->store_value_on_dtor = NULL;
+   bson_free (ptr);
+}
+
+void
+my_value_free_v (void *ptr)
+{
+   my_value_free ((my_value *) (ptr));
+}
+
+static void
+test_simple ()
+{
+   int destroyed_value = 0;
+   mongoc_shared_ptr ptr = MONGOC_SHARED_PTR_NULL;
+
+   ASSERT (mongoc_shared_ptr_is_null (ptr));
+   ptr = mongoc_shared_ptr_create (my_value_new (), my_value_free_v);
+   ASSERT (!mongoc_shared_ptr_is_null (ptr));
+
+   ASSERT_CMPINT (mongoc_shared_ptr_use_count (ptr), ==, 1);
+
+   mongoc_shared_ptr ptr2 = mongoc_shared_ptr_copy (ptr);
+
+   ASSERT (ptr.ptr == ptr2.ptr);
+   ASSERT (ptr._aux == ptr._aux);
+
+   mongoc_shared_ptr valptr_s = mongoc_shared_ptr_copy (ptr);
+   my_value *valptr = valptr_s.ptr;
+   valptr->store_value_on_dtor = &destroyed_value;
+   valptr->value = 133;
+   mongoc_shared_ptr_reset_null (&valptr_s);
+   /* Value hasn't changed yet */
+   ASSERT_CMPINT (destroyed_value, ==, 0);
+
+   /* Now drop the final reference */
```

Comment:
IIUC, `ptr` is not the final reference. `ptr2` still needs to be reset before the destructor is called.